torch_geometric/loader/temporal_dataloader.py

```diff
@@ -0,0 +1,43 @@
+from torch.utils.data import _utils
+from torch.utils.data.dataloader import _BaseDataLoaderIter
+
+from torch_geometric.loader import DataLoader
+
+
+class _SingleProcessTemporalDataLoaderIter(_BaseDataLoaderIter):
+    def __init__(self, loader):
+        super(_SingleProcessTemporalDataLoaderIter, self).__init__(loader)
+        assert self._timeout == 0
+        assert self._num_workers == 0
+
+    def _next_data(self):
+        index = self._next_index()  # may raise StopIteration
+        data = self._dataset[index]
+        if self._pin_memory:
+            data = _utils.pin_memory.pin_memory(data)
+        return data
+
+
+class TemporalDataLoader(DataLoader):
+    r"""A data loader which merges data objects from a
+    :class:`torch_geometric.data.TemporalData` to a mini-batch.
+
+    Args:
+        dataset (TemporalData): The :obj:`temporalData` from which to load
+            the data.
+        batch_size (int, optional): How many samples per batch to load.
+            (default: :obj:`1`)
+        shuffle (bool, optional): If set to :obj:`True`, the data will be
+            reshuffled at every epoch. (default: :obj:`False`)
+        follow_batch (List[str], optional): Creates assignment batch
+            vectors for each key in the list. (default: :obj:`None`)
+        exclude_keys (List[str], optional): Will exclude each key in the
+            list. (default: :obj:`None`)
+        **kwargs (optional): Additional arguments of
+            :class:`torch.utils.data.DataLoader`.
+    """
+    def __iter__(self) -> '_BaseDataLoaderIter':
```

Comment:
Is there any chance we can implement a more efficient version via `slicing`, e.g.:
```python
class TemporalDataLoader(torch.utils.data.DataLoader):
    def __init__(self, data, batch_size, **kwargs):
        self.events_per_batch = batch_size
        arange = range(0, data.num_events, batch_size)
        super().__init__(arange, collate_fn=self.get_batch, batch_size=1, shuffle=False, **kwargs)

    def get_batch(self, batch_idx):
       return data[batch_idx:batch_idx+self.events_per_batch)
```
We should also disallow the usage of `shuffle`, right?

Reply:
@rusty1s, even with some fixes, this code seems to be not working properly. See:
```
tensor([0], device='cuda:0')
Traceback (most recent call last):
  File "tgn.py", line 196, in <module>
    loss = train()
  File "tgn.py", line 121, in train
    src, pos_dst, t, msg = batch.src, batch.dst, batch.t, batch.msg
AttributeError: 'Tensor' object has no attribute 'src'
```
Since you are setting the dataset param of the constructor of the parent class with the `arange`, when iterating in the data loader, it returns the index of the value in the position of the `arange` instead of an object of `TemporalData` as expected.

Btw, I would say the current implementation has a similar performance than using slicing since the iterator ([here](https://github.com/pyg-team/pytorch_geometric/pull/3985/files#diff-5fbb8e45446d16a368919ad46257d4604232cc28066248438771659d26138c37R15)) is getting all the indexes of the batch at once. For instance, for the first batch, The `index` value is:
```
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199]
```
Which is supported by the `TemporalData.__get_item` method and will return a single `TemporalData` object with all the items for the batch.

I've added the `__init__` method in order to disallow the `shuffle` and updated the docs.